### ISSUE TITLE
bool-compare: configurable "any testing" and "equal consistency"

### ISSUE THREAD

**Issue by max-melentyev** [Antonboom/testifylint#14]
Hey, thanks for your work on linter!

I have a proposal to introduce a config option for bool-compare that will allow `.Equal(true, ...)` and `.Equal(false, ...)`. There are two reasons why `.Equal()` can be more appropriate:

1. Testing values of type `any` or `interface{}`:
```go
result := map[string]any
assert.NoError(t, yaml.Unmarshal(yamlStr, &result))
assert.Equal(t, true, result["flag"])
# instead of
assert.True(t, result["flag"].(bool))
```

2. Asserting multiple values of different types:
```go
assert.Equal(t, "test", object.Str)
assert.Equal(t, true, object.Bool)
assert.Equal(t, 123, object.Int)
# instead of
assert.Equal(t, "test", object.Str)
assert.True(t, object.Bool)
assert.Equal(t, 123, object.Int)
```

What do you think?

**Comment by AlekSi** [comment]
Hit both points immediately after enabling that linter. Disabling it until there is an option.

**Comment by Antonboom** [comment]
Hi, guys!
Thank you for report.

How common your cases?
We always have `//nolint:testifylint` option, per-line or function.

1) Why not
```
b, ok := result["flag"].(bool)
require.True(t, ok)
assert.True(t, b)
```
or 
```
assert.Exactly(t, true, result["flag"])
```
?

2) Why this variant is `more appropriate`?

Imho, it is common practise to use the more specified assertions:
```go
spCtx := oteltrace.SpanContextFromContext(ctx2)
assert.True(t, spCtx.TraceID().IsValid())
assert.Equal(t, traceID, spCtx.TraceID().String())
assert.True(t, spCtx.SpanID().IsValid())
assert.Equal(t, spanID, spCtx.SpanID().String())
```
```go
assert.Equal(t, id, id2)
assert.True(t, id.Matches(id2))
```
```go
msg := msgs[0]
s.Equal(lastMsg.ID, msg.ID)
s.Equal(chat, msg.ChatID)
s.Equal(client, msg.AuthorID)
s.Equal("message #0", msg.Body)
s.True(msg.CreatedAt.Equal(lastMsg.CreatedAt))
s.True(msg.IsVisibleForClient)
s.False(msg.IsVisibleForManager)
s.False(msg.IsBlocked)
s.False(msg.IsService)
s.Equal(lastMsg.InitialRequestID, msg.InitialRequestID)
```

---

Regarding 1) – I think maybe just to disable `bool-compare` for any? Does this really need to be configured?

<br>

Regarding 2) – it looks like a style-dependent thing, and I cannot accept this at first glance, because such logic incredibly increases the complexity of the checker implementation. Which is pretty simple now.

**Comment by AlekSi** [comment]
> How common your cases?

So common that I disabled that linter completely for now.

> 1. Why not

I think the first example is more verbose than it should be and the second is confusing (how booleans could be compared _unexactly_?)

> 2. Why this variant is more appropriate?

In your example, code "jumps" left and right:

```
assert.True(t,
assert.Equal(t,
assert.True(t,
assert.Equal(t,
```
vs
```
assert.Equal(t,
assert.Equal(t,
assert.Equal(t,
assert.Equal(t,
```

> Imho, it is common practise to use the more specified assertions

In isolation – yes, but I think it looks worse when I compare some (but not all) `struct` fields one by one.

**Comment by Antonboom** [comment]
@AlekSi, thanks for reply!

`any` is ignored now.

> In your example, code "jumps" left and right:

You can write
```
assert.Equal(t,
assert.Equal(t,
assert.True(t,
assert.True(t,
```
🙂 

But anyway it's really question of project style.
In opposite of this people use specific assertions to highlight these like in `if nil == err` pattern.

Example from FerretDB:
```go
            assert.NoError(t, err)
            assert.Equal(t, tc.msgHeader, msgHeader)
            assert.Equal(t, tc.msgBody, msgBody)
            assert.Zero(t, br.Len(), "not all br bytes were consumed")
            assert.Zero(t, bufr.Buffered(), "not all bufr bytes were consumed")
```
not
```
            assert.NoError(t, err)
            assert.Equal(t, tc.msgHeader, msgHeader)
            assert.Equal(t, tc.msgBody, msgBody)
            assert.Equal(t, 0, br.Len(), "not all br bytes were consumed")
            assert.Equal(t, 0, bufr.Buffered(), "not all bufr bytes were consumed")
``` 

So there are no changes around this, anyway we have `//nolint //testifylint` or the ability to disable a specific checker.
I wouldn't turn off the entire linter.

Thank you!

P.S FerretDB statistics are pretty nice and I discovered some new checks from its tests 👍 
<details>

<summary>Expand...</summary>

```
$ testifylint --enable-all -json ./... | jq '.[] | .[] | .[].category' | sort | uniq -c                 
   8 "error-is-as"
  11 "error-nil"
   2 "expected-actual"
   1 "float-compare"
   1 "go-require"
  12 "len"
  27 "require-error"
```
</details>
